lib/src/spec.rs

```diff
@@ -91,20 +92,35 @@ pub struct ImageReference {
 
 impl ImageReference {
     /// Returns a canonicalized version of this image reference, preferring the digest over the tag if both are present.
-    pub fn canonicalize(self) -> Result<Self, anyhow::Error> {
-        let reference: oci_spec::distribution::Reference = self.image.parse()?;
-
-        if reference.digest().is_some() && reference.tag().is_some() {
-            let registry = reference.registry();
-            let repository = reference.repository();
-            let digest = reference.digest().expect("digest is present");
-            return Ok(ImageReference {
-                image: format!("{registry}/{repository}@{digest}"),
-                ..self
-            });
+    pub fn canonicalize(self) -> Result<Self> {
+        match self.transport.as_str() {
+            "containers-storage" | "registry" | "oci" => {
```

Comment:
The idea here is that `Transport` is already a proper `enum` that we can match on instead of looking at its string forms.

```
match self.transport {
  Transport::ContainerStorage | Transport::Registry | Transport::OciDir
```
etc

Reply:
Ah, I saw ImageReference.transport is of a string type, so didn't realise I could do this

```
pub struct ImageReference {
    /// The container image reference
    pub image: String,
    /// The container image transport
    pub transport: String,
    /// Signature verification type
    #[serde(skip_serializing_if = "Option::is_none")]
    pub signature: Option<ImageSignature>,
}
```

Reply:
Err sorry I was wrong it's the ostree type that's an enum. There's some larger tech debt here because the json schema stuff couldn't apply to types from foreign crates, but we merged ostree-ext so it's more obvious now how to unify these two things.